### PR TITLE
Add option to anonymize ip requests

### LINKED ISSUE #110: GDPR: Storing of IP Addresses / PI
Because we're storing the IP address of a logged in member (not just CMS members), has GDPR compliance been considered? I would have thought we should have a warning somewhere about this new functionality, maybe even just in the changelogs so people can update their privacy policy accordingly...

## PRs
- https://github.com/silverstripe/silverstripe-framework/pull/10844/
- https://github.com/silverstripe/silverstripe-session-manager/pull/161
- https://github.com/silverstripe/silverstripe-session-manager/pull/150

### PR DESCRIPTION
**Pull request by lekoala**
Requires https://github.com/silverstripe/silverstripe-framework/pull/10844
Fixes https://github.com/silverstripe/silverstripe-session-manager/issues/110

### CODE REVIEW

File: src/Models/LoginSession.php [hunk 1]
```diff
@@ -305,4 +306,33 @@ public static function getSessionLifetime(): int
...
+    {
+        $this->LastAccessed = DBDatetime::now()->Rfc2822();
+        $this->IPAddress = $this->getIpFromRequest($request);
+        $this->write();
+    }
+
+    /**
+     * @param HTTPRequest|null $request
+     * @return string
+     */
```
Comment (ChloeSartorelli) [inline src/Models/LoginSession.php]:
```suggestion
```
This information is already in the typehints - this duplication is likely to get outdated and doesn't provide any additional information

File: src/Models/LoginSession.php [hunk 2]
```diff
@@ -305,4 +306,33 @@ public static function getSessionLifetime(): int
 
         return LoginSession::config()->get('default_session_lifetime');
     }
+
+    /**
+     * Update LastAccessed date and IP address
+     *
+     * @param HTTPRequest $request
+     * @return void
```
Comment (ChloeSartorelli) [inline src/Models/LoginSession.php]:
```suggestion
```

File: src/Models/LoginSession.php [hunk 3]
```diff
@@ -305,4 +306,33 @@ public static function getSessionLifetime(): int
...
+     */
+    public function updateLastAccessed(HTTPRequest $request = null)
```
Comment (ChloeSartorelli) [inline src/Models/LoginSession.php]:
```suggestion
    public function updateLastAccessed(?HTTPRequest $request = null): void
```

File: src/Models/LoginSession.php [hunk 4]
```diff
@@ -305,4 +306,33 @@ public static function getSessionLifetime(): int
...
+        $this->LastAccessed = DBDatetime::now()->Rfc2822();
+        $this->IPAddress = $this->getIpFromRequest($request);
+        $this->write();
+    }
+
+    /**
+     * @param HTTPRequest|null $request
+     * @return string
+     */
+    private function getIpFromRequest(HTTPRequest $request = null): string
```
Comment (ChloeSartorelli) [inline src/Models/LoginSession.php]:
```suggestion
    private function getIpFromRequest(?HTTPRequest $request = null): string
```

File: src/Models/LoginSession.php [hunk 5]
```diff
@@ -305,4 +306,33 @@ public static function getSessionLifetime(): int
...
+    {
+        if (!$request) {
+            return '';
+        }
+        $ip = $request->getIP();
+        if (static::config()->anonymize_ip) {
```
Comment (ChloeSartorelli) [inline src/Models/LoginSession.php]:
```suggestion
        if (static::config()->get('anonymize_ip')) {
```
We don't use the config-as-property syntax for fetching configuration in new code.

File: _config/config.yml [hunk 6]
```diff
@@ -13,6 +13,8 @@ SilverStripe\Admin\LeftAndMain:
     - 'silverstripe/session-manager: client/dist/js/bundle.js'
   extra_requirements_css:
     - 'silverstripe/session-manager: client/dist/styles/bundle.css'
+SilverStripe\SessionManager\Models\LoginSession:
+  anonymize_ip: false
```
Comment (ChloeSartorelli) [inline _config/config.yml]:
```suggestion
```
Please add this default to the LoginSession class in PHP instead

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

**Comment by GuySartorelli** [comment]
Thanks for making those changes. Looks like CI is angry about a missing semicolon, can you please fix that up? Otherwise this looks about ready to merge.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Sorry, I should have noticed this earlier - but since this is using a method that's not going to be available until framework 5.1, please bump the constraint in composer.json for framework from `^5` to `^5.1`

Other than that it looks good and works well locally, so once that's done I'll merge both PRs.

**Review by GuySartorelli — APPROVED** [review]

**Pull request merged**